### PR TITLE
FIX: non-existing path or JSON syntax error for `--bids-filter-file` should raise on error.

### PR DESCRIPTION
**Pull request by bpinsard**
## Changes proposed in this pull request

Check if provided bids-filter-file path exists if not raise an error.
This should avoid cases when the user provides file filters and the pipeline silently runs on other set of file. 

## Documentation that should be reviewed

none

### CODE REVIEW

File: fmriprep/cli/parser.py [hunk 1]
```diff
@@ -58,8 +58,11 @@ def _filter_pybids_none_any(dct):
     def _bids_filter(value):
         from json import loads
 
-        if value and Path(value).exists():
-            return loads(Path(value).read_text(), object_hook=_filter_pybids_none_any)
+        if value:
+            if Path(value).exists():
+                return loads(Path(value).read_text(), object_hook=_filter_pybids_none_any)
+            else:
+                raise parser.error(f"Path does not exist: <{value}>.")
```
Comment (mgxd) [inline fmriprep/cli/parser.py:65]: maybe a `FileNotFoundError` would be more helpful here?
```suggestion
            raise FileNotFoundError(f"Path does not exist: <{value}>.")
```

**Review by mgxd — APPROVED** [review]
LGTM - this should probably be backported to the LTS as well. It should be a relatively clean process if you change the target branch to `maint/20.2.x`

File: fmriprep/cli/parser.py (hunk 1 above)
Comment (bpinsard) [reply fmriprep/cli/parser.py:65]: I followed what is done elsewhere in the parser: 
https://github.com/nipreps/fmriprep/blob/962f0b2e26d522d87ed786f12766001aa65e9dfd/fmriprep/cli/parser.py#L20-L24
but I have no problem changing it.

Comment (mgxd) [reply fmriprep/cli/parser.py:65]: sure that works too, though if you are following that example the function should be converted into a partial:

https://github.com/nipreps/fmriprep/blob/c707938e1462e77b1c3a1216a855ceb0aba70845/fmriprep/cli/parser.py#L76-L78

Comment (bpinsard) [reply fmriprep/cli/parser.py:65]: oh right, my mistake. Would there be any advantage of making it a partial for that specific option?

Comment (mgxd) [reply fmriprep/cli/parser.py:65]: i think its necessary to pass in the defined parser

**Comment by effigies** [comment]
Should this target `maint/20.2.x`?

File: fmriprep/cli/parser.py [hunk 2]
```diff
@@ -55,11 +55,14 @@ def _filter_pybids_none_any(dct):
...
+    def _bids_filter(value, parser):
         from json import loads
 
-        if value and Path(value).exists():
-            return loads(Path(value).read_text(), object_hook=_filter_pybids_none_any)
+        if value:
+            if Path(value).exists():
+                return loads(Path(value).read_text(), object_hook=_filter_pybids_none_any)
+            else:
+                raise parser.error(f"Path does not exist: <{value}>.")
```
Comment (effigies) [inline fmriprep/cli/parser.py:65]: Let's follow the lead of `_is_file` and delegate the basic check while adding a more informative error for bad JSON:

```suggestion
        path = _is_file(value, parser)
        try:
            return loads(path.read_text(), object_hook=_filter_pybids_none_any)
        except JSONDecodeError:
            parser.error(f"Invalid JSON file: <{value}>")
```

File: fmriprep/cli/parser.py [hunk 3]
```diff
@@ -55,11 +55,14 @@ def _filter_pybids_none_any(dct):
             for k, v in dct.items()
         }
 
-    def _bids_filter(value):
+    def _bids_filter(value, parser):
         from json import loads
```
Comment (effigies) [inline fmriprep/cli/parser.py:59]:
```suggestion
        from json import loads, JSONDecodeError
```

**Comment by mgxd** [comment]
@bpinsard I think the CI errors have been fixed, you'll need to rebase off the `maint/20.2.x` branch

**Comment by effigies** [comment]
Rebase again?

File: fmriprep/cli/parser.py (hunk 1 above)
Comment (effigies) [reply fmriprep/cli/parser.py:68]: Have we resolved this discussion? Would it be worth adding a test to make sure that it works, or are there tests in the CI that are covering the necessary cases?

Comment (mgxd) [reply fmriprep/cli/parser.py:68]: I think the partial call got lost in the latest rebase

**Comment by effigies** [comment]
@bpinsard I added a test and fixed a bug. This LGTM at this point, feel free to merge when tests pass.

**Pull request merged**